### ISSUE TITLE
Unable to use kraken

### ISSUE THREAD

**Issue by favieres99** [moosylog/exchange_collectors#4]
Hi,

First of all great job! I was trying to use the kraken one, but I'm unable to get my key right, I'm guessing it is 
'apikey' : 'api-key-//mynumbers//'
'secret' : '//private key//'

Any help would be nice,

Thanks!

**Comment by moosylog** [comment]
Hi Gonzalo,

Thanks for the compliment.

You should just use your key/secret

I'm not sure what you mean with //mynumbers//

apikey = your api key
secret = your secret

https://sites.google.com/view/moosyresearch/documentation/exchange-specific-notes#h.b8q8ic7csxfc

*If you have issues:*

enter 'debug' in the third attribute field in CtrXL config
[image: image.png]


Hope this helps,

wkr,

Eelco

**Comment by favieres99** [comment]
Thanks! I ended up finding other working code in the kraken web. I'm with Gemini now, I've manage to get your code working, but I'm having trouble setting paramaters such as timestamp or limit_trades (getting pastTrades right now), where should I place them?

Thanks in advance!

**Comment by moosylog** [comment]
If you share you Gemini code, I can have a look what you should do
different.

**Comment by favieres99** [comment]
// This is your Private request function
function GEM_PrivateRequest(gemrequest) {
  function HMACSHA384HEX(s, secret) { return ToHex(Utilities.computeHmacSignature(Utilities.MacAlgorithm.HMAC_SHA_384, s, secret)).toString(); }
  function ToHex(s) { return s.map(function(byte) { return ('0' + (byte & 0xFF).toString(16)).slice(-2);}).join('');  }
  function stringToBase64(s) { return (Utilities.base64Encode(s)); }
  
  gemrequest.payload = {"request": gemrequest.version+gemrequest.command, "nonce": new Date().getTime() };
  var params = {
    'method'               : gemrequest.method,
    'muteHttpExceptions'   : true,
    'Content-Type'         : "text/plain",
    'headers': {
      "X-GEMINI-APIKEY"    : gemrequest.apikey,
      "X-GEMINI-PAYLOAD"   : stringToBase64(JSON.stringify(gemrequest.payload)),
      "X-GEMINI-SIGNATURE" : HMACSHA384HEX(stringToBase64(JSON.stringify(gemrequest.payload)), gemrequest.secret),
      "Cache-Control"      : "no-cache"
     }
  }
  return  { uri: gemrequest.uri + gemrequest.version + gemrequest.command  , params: params};
}

//This is the object you pass to the request
var gemrequest = {
  'apikey'   : 'account-o4SsTgmotFbGcwo201h6',
  'secret'   : 'TcfT4GNwY7JKHjaA3Cba6V6hAgH',
  'uri'      : 'https://api.gemini.com',
  'version'  : '/v1/',
  'command'  : type,  // type is a value I pass which could be 'balances', 'mytrades', etc, to have a more flexible function
  'method'   : 'post',
  'payload'  : ''
  }; 

My problem is whenever I try to add <'limit_trades=' + 500>, so it shows the last 500 trades done, I don't know how to add it to the private request so it recognises it as a parameter.

// This is how it was done in Kraken, the variable parameters recieved the corresponding in kraken to 'limit_trades=500'
function KAPI_Private(acc_id, endpoint, parameters, cooldown) {
  if (cooldown != 0 && cooldown % 7 == 0){
    Utilities.sleep(42000)
  }
  var api_key = API_Public_Key[acc_id]
  var api_secret = Utilities.base64Decode(API_Private_Key[acc_id])
  var api_path = Utilities.newBlob('/0/private/' + endpoint).getBytes()
  var api_nonce = Number(new Date().getTime()).toFixed(0)*1000;
  var api_post = 'nonce=' + api_nonce + '&' + parameters
  
  var api_sha256 = Utilities.computeDigest(Utilities.DigestAlgorithm.SHA_256, api_nonce + api_post)
  var api_hmac = Utilities.computeHmacSignature(Utilities.MacAlgorithm.HMAC_SHA_512, api_path.concat(api_sha256), api_secret)
  var api_signature = Utilities.base64Encode(api_hmac)
  
  var http_options = {'method':'POST', 'payload':api_post, 'headers':{'API-Key':api_key, 'API-Sign':api_signature}}
  var http_response = UrlFetchApp.fetch('https://api.kraken.com/0/private/' + endpoint, http_options)
  var api_data = http_response.getContentText()
  console.log(api_data);
  return http_response
}

https://docs.gemini.com/rest-api/?python#get-past-trades //here is gemini API info, wich is of no help with this issue

**Comment by favieres99** [comment]
After some long and boring hours I got it right, thanks anyways

**Comment by moosylog** [comment]
I understand you want to add the  "limit_trade" parameter to the request.
[image: image.png]

My code is pretty generic but you should add the limit_trade to the payload
parameter:

gemrequest.payload = {"request":
gemrequest.version+gemrequest.command, "limit_trades"
: 500, "nonce": new Date().getTime() };

I have not tried this but it should work.

Alternatively you can do something like this:

gemrequest.payload = {"request": gemrequest.version+gemrequest.command,
"nonce": new Date().getTime() };
if (your_condition==true) gemrequest.payload.limit_trade = 500

this will add the parameter to the object

Hope this answers your question?

Cheers,
Eelco

**Comment by favieres99** [comment]
Yes! That is what I ended up coming up with, I now have it fully working for a few endpoints. Thanks!
